File: scaffold/src/main/java/top/xuqingquan/web/publics/DefaultDesignUIController.java
package top.xuqingquan.web.publics;

import android.app.Activity;
import android.graphics.Color;
import android.os.Handler;
import android.os.Message;
import android.text.TextUtils;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import com.google.android.material.bottomsheet.BottomSheetDialog;
import com.google.android.material.snackbar.Snackbar;
import top.xuqingquan.utils.Timber;

public class DefaultDesignUIController extends DefaultUIController {

    private BottomSheetDialog mBottomSheetDialog;
    private static final int RECYCLERVIEW_ID = 0x1001;
    private Activity mActivity = null;
    private WebParentLayout mWebParentLayout;
    private LayoutInflater mLayoutInflater;

    @Override
    public void onJsAlert(android.webkit.WebView view, String url, String message) {
        onJsAlertInternal(view, message);
    }

    public void onJsAlert(com.tencent.smtt.sdk.WebView view, String url, String message) {
        onJsAlertInternal(view, message);
    }

    private void onJsAlertInternal(android.webkit.WebView view, String message) {
        Activity mActivity = this.mActivity;
        if (mActivity == null || mActivity.isFinishing()) {
            return;
        }
        try {
            AgentWebUtils.show(view,
                    message,
                    Snackbar.LENGTH_SHORT,
                    Color.WHITE,
                    mActivity.getResources().getColor(android.R.color.black),
                    null,
                    -1,
                    null);
        } catch (Throwable throwable) {
            Timber.e(throwable);
        }
    }

    private void onJsAlertInternal(com.tencent.smtt.sdk.WebView view, String message) {
        Activity mActivity = this.mActivity;
        if (mActivity == null || mActivity.isFinishing()) {
            return;
        }
        try {
            AgentWebUtils.show(view,
                    message,
                    Snackbar.LENGTH_SHORT,
                    Color.WHITE,
                    mActivity.getResources().getColor(android.R.color.black),
                    null,
                    -1,
                    null);
        } catch (Throwable throwable) {
            Timber.e(throwable);
        }
    }

    @Override
    public void onSelectItemsPrompt(android.webkit.WebView view, String url, String[] ways, Handler.Callback callback) {
        showChooserInternal(url, ways, callback);
    }

    @Override
    public void onSelectItemsPrompt(com.tencent.smtt.sdk.WebView view, String url, String[] ways, Handler.Callback callback) {
        showChooserInternal(url, ways, callback);
    }

    private void showChooserInternal(String url, final String[] ways, final Handler.Callback callback) {
        Timber.i("url:" + url + "  ways:" + ways[0]);
        RecyclerView mRecyclerView;
        if (mBottomSheetDialog == null) {
            mBottomSheetDialog = new BottomSheetDialog(mActivity);
            mRecyclerView = new RecyclerView(mActivity);
            mRecyclerView.setLayoutManager(new LinearLayoutManager(mActivity));
            mRecyclerView.setId(RECYCLERVIEW_ID);
            mBottomSheetDialog.setContentView(mRecyclerView);
        }
        mRecyclerView = mBottomSheetDialog.getDelegate().findViewById(RECYCLERVIEW_ID);
        if (mRecyclerView == null) {
            return;
        }
        mRecyclerView.setAdapter(getAdapter(ways, callback));
        mBottomSheetDialog.setOnCancelListener(dialog -> {
            if (callback != null) {
                callback.handleMessage(Message.obtain(null, -1));
            }
        });
        mBottomSheetDialog.show();
    }

    private RecyclerView.Adapter getAdapter(final String[] ways, final Handler.Callback callback) {
        return new RecyclerView.Adapter<BottomSheetHolder>() {
            @NonNull
            @Override
            public BottomSheetHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
                return new BottomSheetHolder(mLayoutInflater.inflate(android.R.layout.simple_list_item_1, viewGroup, false));
            }

            @Override
            public void onBindViewHolder(@NonNull BottomSheetHolder bottomSheetHolder, final int i) {
                TypedValue outValue = new TypedValue();
                mActivity.getTheme().resolveAttribute(android.R.attr.selectableItemBackground, outValue, true);
                bottomSheetHolder.mTextView.setBackgroundResource(outValue.resourceId);
                bottomSheetHolder.mTextView.setText(ways[i]);
                bottomSheetHolder.mTextView.setOnClickListener(v -> {
                    if (mBottomSheetDialog != null && mBottomSheetDialog.isShowing()) {
                        mBottomSheetDialog.dismiss();
                    }
                    Message mMessage = Message.obtain();
                    mMessage.what = i;
                    callback.handleMessage(mMessage);
                });
            }

            @Override
            public int getItemCount() {
                return ways.length;
            }
        };
    }

    private static class BottomSheetHolder extends RecyclerView.ViewHolder {
        TextView mTextView;

        public BottomSheetHolder(View itemView) {
            super(itemView);
            mTextView = itemView.findViewById(android.R.id.text1);
        }
    }

    @Override
    public void bindSupportWebParent(WebParentLayout webParentLayout, Activity activity) {
        super.bindSupportWebParent(webParentLayout, activity);
        this.mActivity = activity;
        this.mWebParentLayout = webParentLayout;
        mLayoutInflater = LayoutInflater.from(mActivity);
    }

    @Override
    public void onShowMessage(String message, String from) {
        if (!TextUtils.isEmpty(from) && from.contains("performDownload")) {
            return;
        }
        if (AgentWebConfig.hasX5()) {
            onJsAlertInternal(mWebParentLayout.getX5WebView(), message);
        } else {
            onJsAlertInternal(mWebParentLayout.getWebView(), message);
        }
    }
}

File: scaffold/src/main/java/top/xuqingquan/web/publics/DefaultDesignUIController.kt
package top.xuqingquan.web.publics

import android.app.Activity
import android.graphics.Color
import android.os.Handler
import android.os.Message
import android.text.TextUtils
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.bottomsheet.BottomSheetDialog
import com.google.android.material.snackbar.Snackbar
import top.xuqingquan.utils.Timber
import top.xuqingquan.web.nokernel.WebConfig

class DefaultDesignUIController : DefaultUIController() {

    private var mBottomSheetDialog: BottomSheetDialog? = null
    private var mActivity: Activity? = null
    private var mWebParentLayout: WebParentLayout? = null
    private var mLayoutInflater: LayoutInflater? = null

    override fun onJsAlert(view: android.webkit.WebView, url: String, message: String) {
        onJsAlertInternal(view, message)
    }

    override fun onJsAlert(view: com.tencent.smtt.sdk.WebView, url: String, message: String) {
        onJsAlertInternal(view, message)
    }

    private fun onJsAlertInternal(view: android.webkit.WebView, message: String) {
        if (this.mActivity == null || this.mActivity!!.isFinishing||this.mActivity!!.isDestroyed) {
            return
        }
        try {
            AgentWebUtils.show(
                view,
                message,
                Snackbar.LENGTH_SHORT,
                Color.WHITE,
                ContextCompat.getColor(this.mActivity!!, android.R.color.black),
                null,
                -1, null
            )
        } catch (throwable: Throwable) {
            Timber.e(throwable)
        }

    }

    private fun onJsAlertInternal(view: com.tencent.smtt.sdk.WebView, message: String) {
        if (this.mActivity == null || this.mActivity!!.isFinishing||this.mActivity!!.isDestroyed) {
            return
        }
        try {
            AgentWebUtils.show(
                view,
                message,
                Snackbar.LENGTH_SHORT,
                Color.WHITE,
                ContextCompat.getColor(this.mActivity!!, android.R.color.black),
                null,
                -1, null
            )
        } catch (throwable: Throwable) {
            Timber.e(throwable)
        }

    }

    override fun onSelectItemsPrompt(
        view: android.webkit.WebView,
        url: String,
        ways: Array<String>,
        callback: Handler.Callback
    ) {
        showChooserInternal(url, ways, callback)
    }

    override fun onSelectItemsPrompt(
        view: com.tencent.smtt.sdk.WebView,
        url: String,
        ways: Array<String>,
        callback: Handler.Callback
    ) {
        showChooserInternal(url, ways, callback)
    }

    private fun showChooserInternal(url: String, ways: Array<String>, callback: Handler.Callback?) {
        if (this.mActivity == null || this.mActivity!!.isFinishing||this.mActivity!!.isDestroyed) {
            return
        }
        Timber.i("url:" + url + "  ways:" + ways[0])
        var mRecyclerView: RecyclerView?
        if (mBottomSheetDialog == null) {
            mBottomSheetDialog = BottomSheetDialog(mActivity!!)
            mRecyclerView = RecyclerView(mActivity!!)
            mRecyclerView.layoutManager = LinearLayoutManager(mActivity)
            mRecyclerView.id = RECYCLERVIEW_ID
            mBottomSheetDialog!!.setContentView(mRecyclerView)
        }
        mRecyclerView = mBottomSheetDialog!!.delegate.findViewById(RECYCLERVIEW_ID)
        mRecyclerView?.adapter = getAdapter(ways, callback)
        mBottomSheetDialog!!.setOnCancelListener {
            callback?.handleMessage(Message.obtain(null, -1))
        }
        mBottomSheetDialog!!.show()
    }

    private fun getAdapter(ways: Array<String>, callback: Handler.Callback?): RecyclerView.Adapter<*> {
        return object : RecyclerView.Adapter<BottomSheetHolder>() {
            override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): BottomSheetHolder {
                return BottomSheetHolder(
                    mLayoutInflater!!.inflate(
                        android.R.layout.simple_list_item_1,
                        viewGroup,
                        false
                    )
                )
            }

            override fun onBindViewHolder(bottomSheetHolder: BottomSheetHolder, i: Int) {
                val outValue = TypedValue()
                mActivity!!.theme.resolveAttribute(android.R.attr.selectableItemBackground, outValue, true)
                bottomSheetHolder.mTextView.setBackgroundResource(outValue.resourceId)
                bottomSheetHolder.mTextView.text = ways[i]
                bottomSheetHolder.mTextView.setOnClickListener {
                    if (mBottomSheetDialog != null && mBottomSheetDialog!!.isShowing) {
                        mBottomSheetDialog!!.dismiss()
                    }
                    val mMessage = Message.obtain()
                    mMessage.what = i
                    callback!!.handleMessage(mMessage)
                }
            }

            override fun getItemCount(): Int {
                return ways.size
            }
        }
    }

    private class BottomSheetHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        internal var mTextView: TextView = itemView.findViewById(android.R.id.text1)
    }

    override fun bindSupportWebParent(webParentLayout: WebParentLayout, activity: Activity) {
        super.bindSupportWebParent(webParentLayout, activity)
        this.mActivity = activity
        this.mWebParentLayout = webParentLayout
        mLayoutInflater = LayoutInflater.from(mActivity)
    }

    override fun onShowMessage(message: String, intent: String) {
        if (this.mActivity == null || this.mActivity!!.isFinishing||this.mActivity!!.isDestroyed) {
            return
        }
        if (!TextUtils.isEmpty(intent) && intent.contains("performDownload")) {
            return
        }
        if (WebConfig.hasX5()) {
            onJsAlertInternal(mWebParentLayout!!.x5WebView!!, message)
        } else {
            onJsAlertInternal(mWebParentLayout!!.webView!!, message)
        }
    }

    companion object {
        private const val RECYCLERVIEW_ID = 0x1001
    }
}
